### ISSUE TITLE
Request for clarification on how onActive()/onInactive() relates to draw settings plus projects

### ISSUE THREAD

**Issue by ssilverman** [heronarts/LXStudio#5]
I'm a little confused as to where `onActive()`/`onInactive()` occur. I'm discovering that things happen in this order:

1. Constructor.
2. `onActive()`.
3. Set any properties from the project file.
4. Bunch of `draw()` calls.
5. `onInactive()`.

I was expecting that `onActive()` and `onInactive()` surround `draw()` calls, and I've been treating `onActive()` as the equivalent to setup() for a processing sketch. Is it true that setting properties from a project file happens _after_ `onActive()`?

**Comment by mcslee** [comment]
A channel can hold multiple patterns, which are transitioned between. onActive() and onInactive() will be invoked each time a pattern is selected on a channel. Note that there are also onTransitionStart() and onTransitionEnd() methods that will be called if a transition is taking place.

Relevant code is here:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L609
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L629

What you're seeing in terms of the order above is slightly strange for a pattern being loaded from a project file. You can see that loading code here - the patterns parameters should be loaded before the pattern is added to the channel:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L765

Are you initializing the pattern in code? I recommend throwing a "new Exception().printStackTrace()" inside the onActive() method to see where this call is coming from?

There are a few other spots in LXChannel that call onActive() in response to programmatic calls. Those are:

(1) Programmatic call to setPatterns(): https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L352

(2) Special case where the active pattern is removed:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L429

(3) Change from one pattern to another:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L615

(4) At the end of loading a channel:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L788

It's not obvious to me why any of these would happen prior to the project file unless you are adding some patterns programmatically via code. I think printing the stack-trace will be the easiest way to track down the source of the call.

**Comment by ssilverman** [comment]
All I know is that I'm seeing this:
1. Set a property in onActive().
2. Notice it's different in draw().

I'm specifically setting it for the lx.engine.audio.output.looping property (and some others). Because that seems to be a system-wide component, maybe there are some things going on between `onActive()` and `draw()`? Deleting "looping":false from the project file I notice doesn't help.

I solved this by adding a protected `setup()` call that's called once before all draw() calls in our LXPattern subclass (and parent to all patterns). (It's the Rainbow Bridge project if you want to have a look. I have this feeling that something's wonky in our project construction. I joined later and didn't write most of the LX-using system, but I'm trying to suss out how patterns interact with everything, and method ordering and unexpected property values have been ongoing issues for me.)

**Comment by mcslee** [comment]
Ah, I see what's going on here. Yes, so the onActive() method is *specifically* for an LXPattern. It means that this particular instance of this particular pattern is going to become active and needs to be ready to receive run() calls.

The Audio Engine output looping property is not part of a pattern. Patterns are free to programmatically change that, but that's slightly discouraged as there is absolutely no guarantees that some other code won't change it for some other reason.

I'll just call out that you're referring to draw() - which is the top-level Processing UI loop... that's really a separate thing from what onActive()/onInactive() have to do with. onActive()/onInactive() are purely about individual LXPattern instances, which have a run() method, but no draw().

> Deleting "looping":false from the project file I notice doesn't help.

It will just be re-saved or defaulted.

> I solved this by adding a protected setup() call that's called once before all draw() calls in our LXPattern subclass (and parent to all patterns)

This does not quite sound like the right fix for whatever problem you're having. You really shouldn't have global initializations inside of LXPatterns. An LXPattern is a single animation routine - it ought only to contain code relating to the particular animation it is generating.

If you want to do some global initialization, which it sounds like you want to happen *after* the project file is loaded, you should put it here, after LXStudio is constructed:
https://github.com/tracyscott/RainbowStudio/blob/master/src/main/java/com/giantrainbow/RainbowStudio.java#L170

The LXStudio constructor will:
- Set up the LX framework
- Invoke the initialize() callback - https://github.com/tracyscott/RainbowStudio/blob/master/src/main/java/com/giantrainbow/RainbowStudio.java#L388
- Invoke the onUIReady() callback after building the UI - https://github.com/tracyscott/RainbowStudio/blob/master/src/main/java/com/giantrainbow/RainbowStudio.java#L397
- Load the last project file, which is stored in the file ".lxproject"

So, global initializations that happen *before* project-file loading should go in the "initialize()" callback, and those that happen *after* project-file loading should go after the LXStudio() construction is done.

The reason why your pattern setup code is getting clobbered is because when a project file is loaded, the audio settings are re-loaded from the file AFTER channels + patterns: https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXEngine.java#L1670

Hope this helps explain. If you give some more context on exactly what you're trying to initialize here and when you want it to happen, I can make more specific recommendations about where to put the code.

**Comment by ssilverman** [comment]
I'll clarify: I should have said `run(double deltaMs)`, which PGBase then uses to potentially slow down calls to a protected draw(double deltaMs), which the sub-patterns override. This is how they're doing slower frame rates. (Not my design, but it's what I have.)

**Comment by ssilverman** [comment]
As per your question, I want audio to start when a pattern starts and end when the pattern ends. See [NyanCat](https://github.com/tracyscott/RainbowStudio/blob/master/src/main/java/com/giantrainbow/patterns/NyanCat.java) for an example of what I'm trying to do.

**Comment by ssilverman** [comment]
I'll add: Sometimes it's important for patterns to reset some things when it's being displayed and undo some things when it's not being displayed. I was hoping to use `onActive()` and `onInactive()`, but that doesn't work well with audio. So I added that `setup()`, which is called once when run() is called for a pattern, but then I'm not sure how to re-call it without overriding `onInactive()` and then forcing subclasses to call `super.onInactive()`. That, however, isn't a very clean design.

Basically, I'm looking for lifecycle events for patterns. Unless you can suggest a different way of thinking about it? How to start and stop audio when a pattern is shown and then not shown?

**Comment by ssilverman** [comment]
And, by the way, thank you for your quick and thoughtful responses, and also for that video. It helped. I think `lx.engine.audio.output` can be considered a global setting, per that video, and so I might need another way to start and stop audio for each pattern?

**Comment by mcslee** [comment]
Okay cool - this all makes sense. For what you're conceptually describing, onActive() and onInactive() do sound like the right place to put this code, and you've convinced me that I should go fix the LX project loading logic to ensure that the initial onActive() calls are deferred until the entire project is loaded. I'll do that shortly.

So, some considerations about patterns triggering audio:

(1) As long as you limit yourself to the patterns that do this being on one channel, so that only one of them is ever running at a time, you should be okay.

(2) The global audio system is used for playback so that there can be a global equalizer with modulation and shared gain, etc. I'm not 100% if you guys are using this in your project to drive animations based upon audio data. But, just throwing out there, that if all you care about is playing back sound, you *could* just use the simple Processing sound API: https://processing.org/reference/libraries/sound/index.html

This way the patterns won't be fighting with each other (or other components) - and multiple sounds could overlap, etc. But I'm thinking there's a decent chance that the reason you're using the LX audio system is to get the spectrum analysis.

(3) Note that if you have transitions enabled on a channel (e.g. a Dissolve between pattern changes) - which I highly recommend so the whole display doesn't just suddenly change to something new - then two patterns can and will overlap. Pattern 1 will go onActive(), then pattern 2 will go onActive(), and the transition will occur for whatever duration is set before pattern 1 goes onInactive(). Something to keep in mind.

There's one other architecture I might suggest here - which is that LX enables you to add anything you like to the global run loop via LXLoopTask: http://lx.studio/api/heronarts/lx/LXLoopTask.html

One thing you could do is something like this:
```
class RainbowAudioWatchdog implements LXLoopTask {

  private LXPattern activePattern = null;
  private InputStream inputStream = null;
  private File file = null;
  private boolean needsUpdate = false;

  public void loop(double deltaMs) {
    // This will be called every iteration of the run-loop, you could check that the
    // correct audio is playing and update it when needed
    if (this.needsUpdate) {
      // Copy over to the temp file, update the audio engine, etc.
      this.needsUpdate = false;
    }
  }

  public void requestPlaybackStart(LXPattern pattern, InputStream stream) {
    // You could store state in this watchdog class about what's *supposed* to be playing now
    if (this.activePattern != pattern && this.inputStream != stream) {
      this.activePattern = pattern;
      this.inputStream = stream;
      this.needsUpdate = true;
    }
  }

  public void requestPlaybackEnd(LXPattern pattern) {
    // You could handle the transition overlap case here... e.g. pattern 1 says "i'm done playing" -
    // but pattern 2 has already requested something new
    if (this.activePattern == pattern) {
      this.inputStream = null;
    }
  }
}

RainbowAudioWatchdog audioWatchdog = new RainbowAudioWatchdog();
lx.engine.addLoopTask(audioWatchdog);
```

This might be unnecessary cruft with the load-order change - but having a little middleman-registry like this would enable you to track the state of which patterns are requesting what audio, and you could handle some conflicts there as well as keep an eye on the audio engine.

**Comment by mcslee** [comment]
FYI: https://github.com/heronarts/LX/commit/8e66521611384181bfd19c143225b8f5df2a73c5

If you pull that update into your project, then I *think* you'll find that onActive() and onInactive() start doing what you want (keeping in mind that if you have transitions active on the channel, then they are going to be interleaved and you may also need to use the onTransitionBegin() and onTransitionEnd() methods).

**Comment by mcslee** [comment]
> And, by the way, thank you for your quick and thoughtful responses, and also for that video. It helped.

Sure thing! I appreciate seeing the framework used, and it's definitely helpful for me to receive questions like this to see where people run into issues trying to do things that I haven't necessarily anticipated. My goal is for it to be as braindead-as-possible for folks to get up and running with this, and I generally try to optimize around that - but that can definitely yield some gotchas when trying to stretch the framework (e.g. this type of tight-coupling animation + audio situation).

Keep the feedback coming!

**Comment by ssilverman** [comment]
1. That's a great idea having a task on the global thread handling the global sound resource. It's obvious when you point it out; why didn't I think of that? It's a standard global resource management technique. :)
2. All I care about for now is playing sound. Using the Processing framework is a decent idea. Do you know if playing multiple things will automatically mix? My hunch is yes, because it's probably using the computer's underlying sound system. I was using the LX audio system because it's there and convenient. I didn't realize its use isn't the same as "play a sound" and "stop a sound". Now that you explained that in the GUI, LXAudioOutput is really meant as just a global resource, it makes more sense.
3. I was spending too much time figuring out how to use the FourierTransform "Bands" for what I needed (beat detection), so I'm using the ddf.minim library's FFT to do my own beat detect (actually using Minim's beat detect because it was even easier than adding my own code, for now). We're also using beat detection for inputs to patterns.
4. Any chance of either:
    1. Getting the changes into a latest build of LXStudio/P3LX? Since nothing is versioned, let alone set up in the Maven global repository, it would be easier to manage the libraries if we didn't have to add.
    2. Adding versions to LXStudio and/or LX and/or building the new code into a versioned LXStudio.
    3. Adding LXStudio and LX to [search.maven.org](search.maven.org) so we didn't have to manually add; the Maven system will take care of getting the latest.

**Comment by mcslee** [comment]
> All I care about for now is playing sound. Using the Processing framework is a decent idea. Do you know if playing multiple things will automatically mix?

Not 100% sure but I suspect they will.

> I was spending too much time figuring out how to use the FourierTransform "Bands" for what I needed (beat detection)

If you want, you can do this 100% in the UI, mapping EQ bands to parameters:
https://www.dropbox.com/s/qhl735o5pyhjpr2/LX%20EQ%20Mapping.mov?dl=0

Alternately, if you're using code, you shouldn't really reach into the FFT class yourself.


```
// If you want your own one, put this in your pattern initialization
GraphicMeter eq = (GraphicMeter) startModulator(lx.engine.audio.input, 16);

// Then, somewhere inside your run method...
for (int i = 0; i < eq.numBands; ++i) {
  // Normalized values (0-1) for each of the bands (not FFT-land, but in user-friendly 16-bands)
  double band = eq.bands[i].getValue();
}

// Want to just use the global LX Meter?
GraphicMeter meter = lx.engine.audio.meter;
```
Rolling your own with Minim is fine too...

> Any chance of either: Getting the changes into a latest build of LXStudio/P3LX?

I have just published GitHub releases for all the projects:
https://github.com/heronarts/LXStudio/releases/tag/0.1.0-alpha
https://github.com/heronarts/LX/releases/tag/0.1.0-alpha
https://github.com/heronarts/P3LX/releases/tag/0.1.0-alpha

Will that work for you?

> Can you comment on the ordering of onActive()/onInactive()/onTransitionBegin()/onTransitionEnd()

For the pattern that is beginning:
(1) onActive() for the pattern that's coming in happens first
(2) ONLY IF transitions are activated, onTransitionBegin() for the new pattern
(3) onInactive() is called for the pattern that's finishing
(4) ONLY IF transitions are activated, onTransitionEnd() happens for the old pattern

The relevant code is all here in the startTransition() and finishTransition() methods on LXChannel:
https://github.com/heronarts/LX/blob/master/src/heronarts/lx/LXChannel.java#L609

> Are onTransitionBegin()/onTransitionEnd() calls made even when there are no transitions? If so, I could just standardize on these instead.

They are not, though I am open to changing that. But onActive() / onInactive() would be the better One True Way for lifecycle start/stop.

**Comment by mcslee** [comment]
Closing this out... feel free to re-open or file a new one if you're still encountering any issues or lack of clarity relating to this.
